simulator/src/testbench.js

```diff
@@ -1,3 +1,512 @@
+import Scope from './circuit';
+import { showError } from './utils';
+
+/**
+ * @typedef {number} RunContext
+ */
+const CONTEXT = {
+    CONTEXT_SIMULATOR: 0,
+    CONTEXT_ASSIGNMENTS: 1,
+};
+
+/**
+ * @typedef {number} Mode
+ */
+const MODE = {
+    MODE_RUNALL: 0,
+    MODE_MANUAL: 1,
+};
+
+// Do we have any other function to do this?
+// Utility function. Converts decimal number to binary string
+function dec2bin(dec) {
+    return (dec >>> 0).toString(2);
+}
+
+/**
+ * UI Function
+ * Create prompt for the testbench UI
+ */
+export function createTestBenchPrompt() {
+    scheduleBackup();
+    const s = `
+    <p>Enter the Test JSON: <input id='testJSON' type='text'  placeHolder='{"type": "comb", ...}'></p>
+    `;
+    $('#testBench').dialog({
+        resizable: false,
+        width: 'auto',
+        buttons: [
+            {
+                text: 'Run All',
+                click() {
+                    const testJSON = $('#testJSON').val();
+                    runTestBench(JSON.parse(testJSON), globalScope, MODE.MODE_RUNALL, CONTEXT.CONTEXT_SIMULATOR);
+                },
+            },
+
+            {
+                text: 'Run Manually',
+                click() {
+                    const testJSON = $('#testJSON').val();
+                    $(this).dialog('close');
+                    runTestBench(JSON.parse(testJSON), globalScope, MODE.MODE_MANUAL, CONTEXT.CONTEXT_SIMULATOR);
+                },
+            },
+        ],
+    });
+
+    $('#testBench').empty();
+    $('#testBench').append(s);
+}
+
+/**
+ * Interface function to run testbench. Called by testbench prompt on simulator or assignments
+ * @param {Object} data - Object containing Test Data
+ * @param {Mode=} mode - To run all tests automatically or let user manually control tests
+ * @param {RunContext=} runContext - Whether simulator or Assignment called this function
+ * @param {Scope=} scope - the circuit
+ */
+export function runTestBench(data, scope = globalScope, mode = MODE.MODE_RUNALL, runContext = CONTEXT.CONTEXT_SIMULATOR) {
+    if (mode === MODE.MODE_MANUAL && runContext === CONTEXT.CONTEXT_ASSIGNMENTS) {
+        // Assignment does not support MANUAL mode
+        return;
+    }
+    const isValidData = validate(data, scope);
+    if (!isValidData.ok){
+        showError(`TestBench: ${isValidData.message}`);
+        return;
+    }
+
+    let results;
+    if (mode === MODE.MODE_RUNALL) results = runAll(data, scope);
+    else if (mode === MODE.MODE_MANUAL) {
+        runManual(data, scope);
+        return;
+    } else {
+        showError('TestBench: Bad Test Mode');
+        return;
+    }
+
+    console.log(results);
+}
+
+/**
+ * Run all the tests automatically. Called by runTestBench()
+ * @param {Object} data - Object containing Test Data
+ * @param {Scope=} scope - the circuit
+ */
+function runAll(data, scope) {
+    // Stop the clocks
+    // TestBench will now take over clock toggling
+    changeClockEnable(false);
```

Comment:
Actually disabling clock is not needed as this is a blocking function but it is still nice to keep if the behavior changes in the future. But you missed an important thing. You have to make the initial clock value to 0.

Reply:
Since we do not step the clock halfway for now, it shouldn't matter if we set the clock to 0 right because one of the two ticks will always be an up edge